## TODO: Determine if sample_metadataframe_df or sample_df should be used
## Retrieve FASTQ directories for spaceranger count input
rule retrieve_fastqs_dir:
    output:
        fastqs_dir=directory(paths.input.fastqs_dir)
    benchmark:
        "benchmark/{sample}_retrieve_fastqs_dir.tab"
    log:
        "log/{sample}_retrieve_fastqs_dir.log"
    params:
        fastqs_dir_uri=lambda wildcards: sample_metadata_df.loc[sample_metadata_df["samid"]==wildcards.sample, "fastqs_dir"].values[0]+"/*",
        cloud=CLOUD,
    shell:
        '''
          echo "mkdir {output.fastqs_dir} && {params.cloud} cp -r {params.fastqs_dir_uri} {output.fastqs_dir}" | tee {log}
          mkdir {output.fastqs_dir} && {params.cloud} cp -r {params.fastqs_dir_uri} {output.fastqs_dir} 2>> {log}
        '''

## Retrieve image for spaceranger count parameter cytaimage.
## Not every specimen will have a brightfield image generated by the CytAssist instrument (--cytaimage)
rule retrieve_cytaimage:
    output:
        done=paths.input.cytaimage
    benchmark:
        "benchmark/{sample}_retrieve_cytaimage.tab"
    log:
        "log/{sample}_retrieve_cytaimage.log"
    params:
        cytaimage_uri=lambda wildcards: image_path(wildcards.sample, "cytaimage")[0],
        cytaimage=lambda wildcards: image_path(wildcards.sample, "cytaimage")[1],
        cloud=CLOUD
    shell:
        '''
          echo "{params.cloud} cp {params.cytaimage_uri} {params.cytaimage} && touch {output.done}" | tee {log}
          {params.cloud} cp {params.cytaimage_uri} {params.cytaimage} && touch {output.done} 2>> {log}
        '''

## Retrieve image for spaceranger count parameter image.
## Not every specimen will have a brightfield microscope image (--image)
rule retrieve_image:
    output:
        done=paths.input.image
    benchmark:
        "benchmark/{sample}_retrieve_image.tab"
    log:
        'log/{sample}_retrieve_image.log'
    params:
        image_uri=lambda wildcards: image_path(wildcards.sample, "image")[0],
        image=lambda wildcards: image_path(wildcards.sample, "image")[1],
        cloud=CLOUD
    shell:
        '''
          echo "{params.cloud} cp {params.image_uri} {params.image} && touch {output.done}" | tee {log}
          {params.cloud} cp {params.image_uri} {params.image}  && touch {output.done} 2>> {log}
        '''

## Retrieve image for spaceranger count parameter darkimage.
## Not every specimen will have a dark background fluorescence microscope image (--darkimage)
rule retrieve_darkimage:
    output:
        done=paths.input.darkimage
    benchmark:
        "benchmark/{sample}_retrieve_darkimage.tab"
    log:
        "log/{sample}_retrieve_darkimage.log"
    params:
        darkimage_uri=lambda wildcards: image_path(wildcards.sample, "darkimage")[0],
        darkimage=lambda wildcards: image_path(wildcards.sample, "darkimage")[1],
        cloud=CLOUD
    shell:
        '''
          echo "{params.cloud} cp {params.darkimage_uri} {params.darkimage} && touch {output.done}" | tee {log}
          {params.cloud} cp {params.darkimage_uri} {params.darkimage} && touch {output.done} 2>> {log}
        '''
## Retrieve image for spaceranger count parameter colorizedimage.
## Not every specimen will have a composite colored fluorescence microscope image (--colorizedimage)
rule retrieve_colorizedimage:
    output:
        done=paths.input.colorizedimage
    benchmark:
        "benchmark/{sample}_retrieve_colorizedimage.tab"
    log:
        "log/{sample}_retrieve_colorizedimage.log"
    params:
        colorizedimage_uri=lambda wildcards: image_path(wildcards.sample, "colorizedimage")[0],
        colorizedimage=lambda wildcards: image_path(wildcards.sample, "colorizedimage")[1],
        cloud=CLOUD
    shell:
        '''
          echo "{params.cloud} cp {params.colorizedimage_uri} {params.colorizedimage}" | tee {log}
          {params.cloud} cp {params.colorizedimage_uri} {params.colorizedimage} && touch {output.done} 2>> {log}
        '''

## Retrieve loupe alignment JSON for spaceranger count parameter loupe-alignment
rule retrieve_loupe_alignment:
    output:
        loupe_json=paths.input.loupe_json
    benchmark:
        "benchmark/{sample}_retrieve_loupe_alignment.tab"
    log:
        "log/{sample}_retrieve_loupe_alignment.log"
    params:
        loupe_json_uri=lambda wildcards: image_path(wildcards.sample, "loupe_alignment"),
        cloud=CLOUD
    shell:
        '''
          echo "{params.cloud} cp {params.loupe_json_uri} {output.loupe_json}" | tee {log}
          {params.cloud} cp {params.loupe_json_uri} {output.loupe_json} 2>> {log}
        '''

## Determine if the multiple flow cells were used for sequencing.
## Generates the list needed for the --sample parameter in spaceranger count
rule generate_sample_param:
    input:
        fastqs_dir=rules.retrieve_fastqs_dir.output.fastqs_dir
    output:
        sample_param=paths.input.sample_param
    benchmark:
        "benchmark/{sample}_generate_sample_param.tab"
    log:
        "log/{sample}_generate_sample_param.log"
    params:
        srcdir=SOURCEDIR+"/python/generate-sample-param.py",
        sample="{sample}"
    script:
          "../source/python/generate-sample-param.py"
